### PR TITLE
[Metadata management] Query Translation & Query Runner

### PR DESCRIPTION
**Pull request by josephjacobmorris**
**What type of PR is this?**
> /kind new feature

**What this PR does / why we need it**:
This PR implements the Query Translator and Query Runner according to design analysis
Fixes #

**Test Report Added?**:
> /kind TESTED

**Test Report**:
* No Params
image

* When backendName is specified
1.valid backend name
image
2. Invalid backend name
image

* Based on region
1. Valid
image
2. Invaild
image

* When BucketName is specified
1. Valid
image
* When filtered based on bucket size
image
* When Filter via object name
1.Valid
image
2.Invalid
image

* When filter via object size
 Valid
image
**Special notes for your reviewer**:

### CODE REVIEW

File: metadata/pkg/constants/Constants.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+		"ap-northeast-1", "ca-central-1", "eu-central-1", "eu-west-1", "eu-west-2", "eu-south-1", "eu-west-3",
+		"eu-south-2", "eu-north-1", "eu-central-2", "me-south-1", "me-central-1", "sa-east-1", "us-gov-east-1",
+		"us-gov-west-1"}
+)
+
+const (
+	ZERO          = 0
+	DOLLAR_SYMBOL = "$"
+	DOLLAR_DOLLAR = "$$"
+
```
Comment (josephjacobmorris) [inline metadata/pkg/constants/Constants.go:28]: remove space

File: metadata/pkg/constants/Constants.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	//comparioson operators
+	DOLLAR_EQUAL_OPERATOR = "$eq"
+	//logical_operators
+	OR_OPERATOR  = "$or"
+	AND_OPERATOR = "$and"
+
+	MAP_AGG_OPERATOR     = "$map"
+	PROJECT_AGG_OPERATOR = "$project"
+	INCLUDE_FIELD        = 1
+
```
Comment (josephjacobmorris) [inline metadata/pkg/constants/Constants.go:60]: remove space

File: metadata/pkg/query-translator/QueryTranslator.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+)
+
+func Translate(in *pb.ListMetadataRequest) []bson.D {
+	aggOperations := constructAggOperationForBackendLevel(in, make([]bson.D, 0))
+	aggOperations = constructAggOperationForBucketLevel(in, aggOperations)
+	aggOperations = constructAggOperationForObjectLevel2(in, aggOperations)
+	return aggOperations
+}
+
+func constructAggOperationForObjectLevel2(in *pb.ListMetadataRequest, aggOperations []bson.D) []bson.D {
```
Comment (josephjacobmorris) [inline metadata/pkg/query-translator/QueryTranslator.go:17]: rename function to constructAggOperationForObjectLevel2

File: metadata/pkg/query-translator/QueryTranslator.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+import (
+	"github.com/opensds/multi-cloud/metadata/pkg/constants"
+	pb "github.com/opensds/multi-cloud/metadata/proto"
+	log "github.com/sirupsen/logrus"
+	"go.mongodb.org/mongo-driver/bson"
+)
+
+func Translate(in *pb.ListMetadataRequest) []bson.D {
+	aggOperations := constructAggOperationForBackendLevel(in, make([]bson.D, 0))
```
Comment (josephjacobmorris) [inline metadata/pkg/query-translator/QueryTranslator.go:11]: make([]bson.D, 0) with var aggOperations []bson.D

Comment (josephjacobmorris) [reply metadata/pkg/query-translator/QueryTranslator.go]: done

File: metadata/pkg/constants/Constants.go (hunk 2 above)
Comment (josephjacobmorris) [reply metadata/pkg/constants/Constants.go]: done

File: metadata/pkg/constants/Constants.go (hunk 1 above)
Comment (josephjacobmorris) [reply metadata/pkg/constants/Constants.go]: done

**Review by josephjacobmorris — COMMENTED** [review]
Self-Review done

**Review by PravinRanjan10 — CHANGES_REQUESTED** [review]

File: api/pkg/metadata/service.go [hunk 5]
```diff
@@ -94,12 +95,12 @@ func GetListMetaDataRequest(request *restful.Request) (listMetadataRequest mt.Li
 		return mt.ListMetadataRequest{}, err
 	}
 
-	bucketSizeOperator := request.PathParameter("BucketSizeOperator")
-	objectSizeOperator := request.PathParameter("ObjectSizeOperator")
+	bucketSizeOperator := request.QueryParameter("BucketSizeOperator")
+	objectSizeOperator := request.QueryParameter("ObjectSizeOperator")
 	limit, offset, err := common.GetPaginationParam(request)
```
[the remark this answers is not in the document]
Comment (josephjacobmorris) [reply api/pkg/metadata/service.go:100]: done

File: metadata/pkg/query-runner/QueryRunner.go [hunk 6]
```diff
@@ -0,0 +1,13 @@
+package queryrunner
```
[the remark this answers is not in the document]
Comment (josephjacobmorris) [reply metadata/pkg/query-runner/QueryRunner.go:1]: done

File: api/pkg/metadata/service.go [hunk 7]
```diff
@@ -94,12 +95,12 @@ func GetListMetaDataRequest(request *restful.Request) (listMetadataRequest mt.Li
...
 
 	return mt.ListMetadataRequest{Type: typeOfCloudVendor, BackendName: backendName, Limit: limit, Offset: offset,
 		BucketName: bucketName, ObjectName: objectName, SizeOfObjectInBytes: sizeOfObjectInBytes,
-		SizeOfBucketInBytes: sizeOfBucketInBytes, BucketSizeOperator: bucketSizeOperator, ObjectSizeOperator: objectSizeOperator}, err
+		SizeOfBucketInBytes: sizeOfBucketInBytes, BucketSizeOperator: bucketSizeOperator, ObjectSizeOperator: objectSizeOperator, Region: region}, err
```
[the remark this answers is not in the document]
Comment (josephjacobmorris) [reply api/pkg/metadata/service.go]: done

File: metadata/pkg/db/drivers/mongo/mongo.go [hunk 8]
```diff
@@ -87,53 +88,53 @@ func UpdateContextFilter(ctx context.Context, m bson.M) error {
...
+		log.Infof("the value of in:%s", backend)
 	}
 
 	return nil
 }
 
-func (ad *adapter) ListMetadata(ctx context.Context, limit int32) ([]*model.MetaBucket, error) {
-	log.Infoln("I am in GetBucket..........")
+func (ad *adapter) ListMetadata(ctx context.Context, query []bson2.D) ([]*model.MetaBackend, error) {
+	log.Infoln("i am in get bucket..........")
```
[the remark this answers is not in the document]
Comment (josephjacobmorris) [reply metadata/pkg/db/drivers/mongo/mongo.go]: done

File: metadata/pkg/query-translator/QueryTranslator.go [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+			{constants.BACKEND_NAME, constants.INCLUDE_FIELD},
+			{constants.REGION, constants.INCLUDE_FIELD},
+			{constants.TYPE, constants.INCLUDE_FIELD},
+			{constants.BUCKETS, bson.D{
+				{constants.MAP_AGG_OPERATOR, bson.D{
+					{constants.INPUT, constants.DOLLAR_SYMBOL + constants.BUCKETS},
+					{constants.AS, constants.BUCKET},
+					{constants.IN, bson.D{
+						//* asking to include all the rest of bucket fields in the map output
+						{constants.CREATION_DATE, constants.DOLLAR_DOLLAR + constants.BUCKET + constants.DOT + constants.CREATION_DATE},
```
[the remark this answers is not in the document]
Comment (josephjacobmorris) [reply metadata/pkg/query-translator/QueryTranslator.go]: done as per discussion , extracted multiple functions to increase readability

[the remark this answers is not in the document]
Comment (josephjacobmorris) [reply metadata/pkg/query-translator/QueryTranslator.go]: done as per discussion, extracted multiple functions to increase readability

File: metadata/pkg/query-translator/QueryTranslator.go [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+
+	if len(filterConditions) > 0 {
+		findMatchingDocuments := bson.D{{constants.MATCH_AGG_OP, bson.D{{
+			constants.BUCKETS, bson.D{{
+				constants.ELEMMATCH_AGG_OPERATOR,
+				bson.D(matchConditions),
+			}},
+		}}}}
+
+		log.Debugln("matching documents query for object level.....", findMatchingDocuments)
```
[the remark this answers is not in the document]
Comment (josephjacobmorris) [reply metadata/pkg/query-translator/QueryTranslator.go]: done

File: metadata/pkg/db/drivers/mongo/mongo.go [hunk 11]
```diff
@@ -87,53 +88,53 @@ func UpdateContextFilter(ctx context.Context, m bson.M) error {
...
 
-	var bucket []*model.MetaBucket
+	log.Infoln("pipeline query..........", pipeline)
 
-	log.Infof("ListMetadata, limit=%d, offset=%d, m=%+v\n", limit, offset, m)
+	database := session.Database(MetadataDataBaseName)
+	collection := database.Collection(MetadataCollectionName)
 
-	cur, err := session.Database(DataBaseName).Collection(BucketMD).Find(ctx, m, options.Find().SetSkip(int64(offset)).SetLimit(int64(limit)))
+	log.Infoln("database..........", database.Name())
```
[the remark this answers is not in the document]
Comment (josephjacobmorris) [reply metadata/pkg/db/drivers/mongo/mongo.go]: done

File: metadata/pkg/query-translator/QueryTranslator.go [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+import (
+	"github.com/opensds/multi-cloud/metadata/pkg/constants"
+	pb "github.com/opensds/multi-cloud/metadata/proto"
+	log "github.com/sirupsen/logrus"
+	"go.mongodb.org/mongo-driver/bson"
+)
+
+func Translate(in *pb.ListMetadataRequest) []bson.D {
+	var aggOperations []bson.D
+	aggOperations = constructAggOperationForBackendLevel(in, aggOperations)
```
Comment (PravinRanjan10) [inline metadata/pkg/query-translator/QueryTranslator.go:12]: can you please add example of aggregation as comment.

File: metadata/pkg/service/service.go [hunk 13]
```diff
@@ -104,29 +107,35 @@ func (f *metadataService) SyncMetadata(ctx context.Context, in *pb.SyncMetadataR
 
 	return nil
 }
-
 func (f *metadataService) ListMetadata(ctx context.Context, in *pb.ListMetadataRequest, out *pb.ListMetadataResponse) error {
-	log.Info("Received GetMetadata request in metadata service.")
-	res, err := db.DbAdapter.ListMetadata(ctx, in.Limit)
-	if err != nil {
-		log.Errorf("Failed to create backend: %v", err)
+	log.Info("received GetMetadata request in metadata service.")
```
Comment (PravinRanjan10) [inline metadata/pkg/service/service.go:111]: better to add info for example:
log.Infof("received GetMetadata request in metadata service:%+v", in)

File: metadata/pkg/utils/ObjectMapper.go [hunk 14]
```diff
@@ -0,0 +1,67 @@
+package utils
+
+import (
+	"github.com/opensds/multi-cloud/metadata/pkg/model"
+	"github.com/opensds/multi-cloud/metadata/proto"
+
+	pb "github.com/opensds/multi-cloud/metadata/proto"
+)
+
+func GetProtoBackends(unPaginatedResult []*model.MetaBackend) []*pb.BackendMetadata {
```
Comment (PravinRanjan10) [inline metadata/pkg/utils/ObjectMapper.go:10]: please rename to simply, GetBackends()

File: metadata/pkg/query-translator/QueryTranslator.go (hunk 12 above)
Comment (josephjacobmorris) [reply metadata/pkg/query-translator/QueryTranslator.go:23]: image
done

File: metadata/pkg/utils/ObjectMapper.go (hunk 14 above)
Comment (josephjacobmorris) [reply metadata/pkg/utils/ObjectMapper.go]: image
done

File: metadata/pkg/service/service.go (hunk 13 above)
Comment (josephjacobmorris) [reply metadata/pkg/service/service.go]: image

done

**Review by PravinRanjan10 — APPROVED** [review]

**Pull request merged**